### PR TITLE
#7956: login banners behind a feature flag (3/3)

### LINKED ISSUE #7955: Slice 1: Create integrationLoginProtocol.ts and implement show/hide banner functions with login button click handler

### LINKED ISSUE #7956: Slice 3: Update contentScriptPlatform.request() to handle UserLoginRequiredError properly and defer response promises

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Closes #7956
- Closes #7955
- Feature flag: Banners are behind `integration-login-banner` feature flag
- Adds "label" to SanitizedIntegrationConfiguration so it can be shown in the UI

## Reviewer Notes

- This PR is stacked on: #7957 - marked as "do not merge" until then. Also needs to be rebased given that branch was rebased off of main
- Review `contentScript/contentScriptPlatform`
- Review `contentScript/integrations/deferredLoginController`
- Review `contentScript/integrations/bannerDomController`
- Review `contentScript/integrations/LoginBanners`
- Review `errorService` and `errors/` updates

## Discussion

- Differences [from the implementation sketch](https://www.notion.so/pixiebrix/Phase-require-user-gesture-before-displaying-Google-and-other-OAuth2-PKCE-popups-when-mods-run-69650e3985ab4d4785b9e1cf3578c9cd?pvs=4#38c47e619a4a4ee2b7287e16c0b0509d):
  - `OAuth2RequestCacheEntry` is not required, as we can just use normal async/await to pause a method while waiting for the deferred login
  -  `onLoginClicked` is not required, as all frames can just listen for changes to the storage

## Remaining Work

- [x] UI for banners: base style, font-size protection
- [x] Add tests
- [x] Adjust error reporting logic
- [x] Record demo

## Demo

- To test, use Storage Area Explorer to clear out the OAuth2 cache
- https://www.loom.com/share/6b1ab13ac405426ca87801d975694aca?sid=50ecfdde-c08d-437f-8f5b-3ad0e03bafa5

## Future Work

- Adjust UI styling based on Brandon's feedback
- Modify the `PlatformProtocol.request` method to accept a modId (and mod name?) so that the banner can display the mod name
- Implement smarter RequestSupersededError strategy: https://github.com/pixiebrix/pixiebrix-extension/blob/f9e3cb7f735e3a4784051285c84db90a9a1dbfc5/src/contentScript/integrations/deferredLoginController.ts#L71-L71

## Checklist

- [x] Add tests
- [x] ~~New files added to `src/tsconfig.strictNullChecks.json` (if possible)~~: can't add because launchInteractiveOAuth2Flow uses integration registry
- [x] Designate a primary reviewer: @BLoe

### CODE REVIEW

File: src/contentScript/contentScriptPlatform.test.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+describe("contentScriptPlatform", () => {
+  it("should use interactive request by default", async () => {
+    appApiMock.onGet("/api/me/").reply(200, { flags: [] });
+
+    const config = sanitizedIntegrationConfigFactory();
+    const request = { url: "https://example.com" };
+    await contentScriptPlatform.request(config, request);
+
+    expect(backgroundRequestMock).toHaveBeenCalledOnce();
+    // `toHaveBeenCalledExactlyOnceWith`'s matching is confusing
```
Comment (grahamlangford) [inline src/contentScript/contentScriptPlatform.test.ts:57]: Curious what you found confusing? We've used it several times in the past.

File: src/contentScript/contentScriptPlatform.test.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      interactiveLogin: false,
+    });
+  });
+
+  it("handles deferred login", async () => {
+    appApiMock
+      .onGet("/api/me/")
+      .reply(200, { flags: ["integration-login-banner"] });
+
+    backgroundRequestMock.mockRejectedValueOnce(
```
Comment (grahamlangford) [inline src/contentScript/contentScriptPlatform.test.ts:104]: We're testing here that the login banner is displayed if the request receives an InteractiveLoginRequiredError, but I don't believe we have any tests anywhere that validate that when `launchWebFlow` throws that error, it isn't swallowed between there and `performConfiguredRequest`.

File: src/contentScript/contentScriptPlatform.test.ts (hunk 1 above)
Comment (twschiller) [reply src/contentScript/contentScriptPlatform.test.ts:57]: I think the matcher feedback is just worse on a failed match (harder to see what doesn't match). I'll switch over to using toHaveBeenCalledExactlyOnceWith for clarity

File: src/contentScript/contentScriptPlatform.test.ts (hunk 2 above)
Comment (twschiller) [reply src/contentScript/contentScriptPlatform.test.ts:110]: Added a test in performConfiguredRequest.test.ts

File: src/background/performConfiguredRequest.test.ts [hunk 3]
```diff
@@ -207,6 +239,48 @@ describe("authenticated direct requests", () => {
...
+      requestConfig,
+      { interactiveLogin: false },
+    );
+
+    await expect(request).rejects.toThrow(ContextError);
+
+    try {
+      await request;
+    } catch (error) {
+      // We're using the itInteractiveLoginRequiredError error type for control flow. Test that it doesn't swallow the
```
Comment (grahamlangford) [inline src/background/performConfiguredRequest.test.ts:263]:
```suggestion
      // We're using the InteractiveLoginRequiredError error type for control flow. Test that it doesn't swallow the
```

**Review by grahamlangford — APPROVED** [review]

File: src/contentScript/contentScriptPlatform.test.ts (hunk 1 above)
Comment (grahamlangford) [reply src/contentScript/contentScriptPlatform.test.ts]: > I think the matcher feedback is just worse on a failed match (harder to see what doesn't match). I'll switch over to using toHaveBeenCalledExactlyOnceWith for clarity

Honestly, that's a good enough reason not to use it for me. I don't know what the errors look like off the top of my head. Though I'd suggest a linting rule to prefer `toHaveBeenCalledOnce` followed by `toHaveBeenCalledWith`.

**Pull request merged**